### ISSUE TITLE
Performance of binding-generator

### ISSUE THREAD

**Issue by jerry73204** [twistedfall/opencv-rust#163]
I'm working on Ubuntu 18.04 with old OpenCV that requires me to enable `buildtime-bindgen` and `clang-runtime` for reasons. I've been suffering from the crappy performance of the binding-generator. I'm looking forward to speed it up.

I noticed that the building routine with `clang-runtime` feature does not employ the job server ([code](https://github.com/twistedfall/opencv-rust/blob/504a4867fbff0e54ea3b91ff25a8f3b8d908a768/build.rs#L113-L129)). Are there reasons that blocks the job server? We would gain the benefit of parallelism if it can be resolved.

The second is that I profiled the elapsed time of each module shown below. The most critical task could be up to 20 seconds. I think we could go beyond the module-wise parallelism, or locale the critical part that consumes lots of time.

```
core 19.944345441s
stitching 8.581334137s
dnn 7.705141037s
imgproc 6.225595582s
rgbd 6.021543039s
calib3d 5.398754344s
videostab 5.053067548s
tracking 4.652928171s
features2d 4.23875516s
ml 3.932232241s
viz 3.930411622s
objdetect 3.80578337s
text 3.668900933s
face 3.64243326s
ximgproc 3.353856684s
optflow 3.107020836s
aruco 3.051257091s
xfeatures2d 3.007160311s
ccalib 2.983264211s
line_descriptor 2.826963486s
video 2.611390679s
photo 2.587911801s
videoio 2.028532128s
saliency 1.975049353s
highgui 1.974326047s
structured_light 1.972092153s
flann 1.946669479s
superres 1.917765315s
hdf 1.909713359s
quality 1.880025585s
surface_matching 1.876563577s
bgsegm 1.873986262s
bioinspired 1.84069713s
stereo 1.725183691s
shape 1.720512943s
rapid 1.671785967s
dnn_superres 1.64045128s
plot 1.46913129s
fuzzy 1.424393002s
imgcodecs 1.420861355s
xphoto 1.397017447s
img_hash 1.360725893s
phase_unwrapping 1.343983614s
xobjdetect 1.326603382s
intensity_transform 1.311640538s
dpm 1.190400088s
hfs 1.174297226s
freetype 1.135215178s
cvv 1.121119644s
alphamat 1.007060339s
```

**Comment by twistedfall** [comment]
Just below the referenced section is the commented out code for the parallel build and it references this issue: https://github.com/twistedfall/opencv-rust/issues/145 

I don’t remember though what was the exact reason for that error message, I’ll need to investigate it.

> I think we could go beyond the module-wise parallelism, or locale the critical part that consumes lots of time.

Yep, the core module is the biggest offender. And it’s the one you can’t avoid :) I would really like to hear your thoughts on how we can parallelize its build.

I do profile the bindings generator from time to time and so far most of the time is spent inside libclang so the only thing we can do here is call it less often. Some kind of caching maybe?

**Comment by jerry73204** [comment]
> I do profile the bindings generator from time to time and so far most of the time is spent inside libclang so the only thing we can do here is call it less often.

I'm curious here. I see the clang indexing ([code](https://github.com/twistedfall/opencv-rust/blob/504a4867fbff0e54ea3b91ff25a8f3b8d908a768/binding-generator/src/generator.rs#L369-L392))  takes up to 4s, while the entity walker could be as long as 15s. My understanding is that clang has done most translation job in that 4s. How would clang involves in entity walker?

I'm thinking of breaking up top level entities (class, enum, func, etc) into parallel tasks. I'm not sure if the visiting order is relevant or not before I know more details. Perhaps you can share some possible challenges.

**Comment by jerry73204** [comment]
I tried `perf record cargo build && perf report` but still cannot observe obvious hotspots. Do you have suggestions to profile the build script?

**Comment by twistedfall** [comment]
I would say to collect sensible performance data you'd need to build the binding generator with `--release` (and debug information). Although I have another question, the binding generation takes a lot of time, but it should only be run once per project (and per rustc upgrade, per crate update), is it not happening for you? Do you see frequent binding regenerations?

**Comment by jerry73204** [comment]
I'm using rust-analyzer in emacs. I noticed the re-compilation whenever I re-open my project in emacs. However, `cargo build` does not invoke recompilation in command line.

**Comment by twistedfall** [comment]
That's probably some specifics of rust-analyzer and/or emacs, not really sure whether it's fixable.

Regarding the parallel `clang-runtime`. To be able to do parallel binding generation the build script compiles a helper binary and runs it in parallel. The problem before was that I couldn't come up with the way to build this helper binary when opencv crate is used as a dependency of another project. The command `cargo build --release --package opencv-binding-generator --bin binding-generator` didn't work because there was no package named `opencv-binding-generator` for the project that just uses opencv crate. But when trying to reproduce it yesterday I couldn't reproduce the issue anymore, so maybe there was some change in cargo that made it work since. I'll try to run some more tests, but in the meanwhile, can you please try using the `parallel-clang-runtime` branch to see if it works for you?

**Comment by jerry73204** [comment]
`rm -rf target && cargo build --release --features clang-runtime` on both stable and nightly-2020-09-21 works fine.

**Comment by twistedfall** [comment]
Parallel build with clang-runtime feature is now available in v0.46.1
